### ISSUE TITLE
SecureNN Integration

### ISSUE THREAD

**Issue by jvmancuso** [mortendahl/tf-encrypted#106]
We (Dropout Labs) are ready to start merging our internal SecureNN prototype into tf-encrypted.

### Integration plan
The plan is for SecureNN to be a separate Protocol class that inherits code from Pond, but switches out Int100Tensor for a new BackingTensor.  As a result, we may not be able to directly subclass Pond as it stands in the code right now, although nearly all of the code will need to be used in the SecureNN Protocol.  More concrete plans and issues for each of these steps are in the works.

- [x] Our first step should be to figure out precisely what we need to do to be able to inherit the Pond Protocol with our new BackingTensor.
- [x] Subclass pond directly and create a SecureNN class.
- [x] Once we've done that, we can write out the new BackingTensor class. i.e. NativeTensor
- [x] Integrate the Select Share protocol (#154)
- [x] Integrate the bit decomposition operation (aka `binarize`) in Tensorflow (using `tf.bitwise.right_shift` or repeated division/modulus as a fallback).  This is required for all further work on SecureNN protocols. Do this on the NativeTensor to start.
- [x] Integrate the bit decomposition operation (aka `binarize`) in Tensorflow for the CrtTensor.
- [x] Integrate the Private Compare protocol (#153)
- [x] Integrate the Share Convert protocol (first as noop for Int100Tensor, then progressively more general for other BackingTensors, see note below)
  - #237
- [x] Integrate the Compute MSB protocol
- [x] Integrate the dReLU & ReLU protocols
- [x] ~Integrate the Division protocol~
  - this is not complete, but will not be done as part of this issue: #258
- [x] Integrate the MaxPool protocol
- [ ] Figure out how to modify which uses of Mul/MatMul need to include the zero masking for hiding results from the crypto provider (conforming to the SecureNN MatMul protocol). (cc @mortendahl)

Here is a convenient reference for each protocol's dependencies.  Squares represent Main Protocols, circles represent Supporting Protocols.  Development can be parallelized accordingly.
<img width="542" alt="SecureNN: Operation Dependency Graph" src="https://user-images.githubusercontent.com/7891333/45457359-92ee5580-b6bc-11e8-903e-b60b827ab565.png">
Note that the bit decomposition operation is not pictured, as it is a prerequisite for Private Compare and is not itself a SecureNN protocol.

### Changes from the original SecureNN paper
Based on our internal experimentation with numpy, we plan on changing a few things from the original paper.
- It's highly likely, pending the result of #100, that we'll be using a BackingTensor based on tf.int64 (as opposed to uint64 from the paper).  This is largely motivated by what's currently/easily enabled in Tensorflow.  For the moment, we'll be developing with Int100Tensor in mind, to avoid complications, although we'll also try to test with planned Int32 and Int64 Tensors by the end.  This is possible because the BackingTensor is abstracted away from Pond, so we can subclass Pond without too much complication.
- As a result, we'll need to switch the ring SecureNN uses from `L = 2 ** 64` according to which BackingTensor we're using.  Again, this is all abstracted away in the BackingTensor implementation.  All results from SecureNN will remain.
- We want to abstract a few parts of SecureNN a bit differently.  In particular, there's a bitwise XOR method that gets used implicitly quite a bit in the protocols that deal with bits.  We'll want that, and any other recurring functionality, abstracted into a standalone protocol method.
- We hope to have our implementation of Share Convert work for converting to arbitrary rings, so that it can be reused in future protocols (should it become useful on its own).  We'll have similar hopes for the rest of the protocols, although this is the most obvious case.
- We'll be using our prototype repo to guide and motivate the integration, to take into account any other adjustments we had to make to the published pseudocode in order to get all protocols working together in numpy.

**Comment by justin1121** [comment]
Notes from discussion with @mortendahl and @jvmancuso:

- Try to abstract certain equations from securenn out into nameable functions. Bit xor in private compare for example.
- Create a native tensor which is represented underneath by an int32 tensor where we can pass any modulus into and do all the mod reductions internally to the class.
- Subclass Pond and create a SecureNN protocol. We will start by using the native tensor in this new protocol.
- Fill in all the needed functions for securenn from top down and then begin filling in the more complicated functions.
- Start work on crt tensor bit extraction.
- Skip the zero masking in matmul for now and just remask shares before sending to the crypto provider.

I will update the top level comment to reflect these decisions.

**Comment by ianlivingstone** [comment]
@mortendahl will update the todo list with our new priorities to complete this issue.

**Comment by mortendahl** [comment]
## Planning from Tuesday meeting

Deadline for using SecureNN benchmark in paper is Friday morning.

First priority:
- [ ] private compare (@bendecoste @justin1121)
- [ ] lsb (@jvmancuso)
- [ ] MaxPool/ReLU (@mortendahl @jvmancuso)

Second priority:
- [ ] Int64 support (@mortendahl @justin1121)
- [ ] share convert (@justin1121)
- [ ] ArgMax

Third priority:
- [ ] Division
- [ ] Softmax

**Comment by bendecoste** [comment]
@mortendahl and I decided the quickest way to have maxpooling working was to essentially copy the flow for averagepooling.

The algorithm from secureNN essentially gives you the max over some axis. ie it does not tell you how you should actually implement maxpooling.  For average pooling we have been using im2col for most cases, and have decided to take this approach for max pooling as well.  We will get the im2col representation of the PondTensor and then call `reduce_max` on that.

I have implemented the im2col representation in the commit above, @mortendahl is implementing `reduce_max`.

**Comment by jvmancuso** [comment]
I don't know if this will extend to SecureNN's `reduce_max`, but the reshape method for average pooling is quite a bit faster there, and the requirements for when it's applicable are pretty common too!

**Comment by ianlivingstone** [comment]
@mortendahl or @bendecoste can you update the checklist at the top of this issue when you have a moment? :)

**Comment by bendecoste** [comment]
# Quick update

The checklist is up to date.  The largest remaining tasks are

- Relu bug
  - Sometimes relu is returning -1.  I believe @mortendahl is looking at this?
- Share convert
  - It runs but returns incorrect results.  E.g. if you share convert [1,2,3,4] and [1,2,3,4] then add them, you will not get [2,4,6,8]

We are considering deferring division into another issue.  If no one objects I will open a new issue tomorrow.

After the above is done then we can close this issue 🎉

**Comment by bendecoste** [comment]
> We are considering deferring division

We are not differing division!

**Comment by ianlivingstone** [comment]
@bendecoste can we get an update to the checklist with the outstanding items and what has been deferred?

Great update!

**Comment by bendecoste** [comment]
> We are considering deferring division into another issue

> can we get an update to the checklist with the outstanding items and what has been deferred?

https://github.com/mortendahl/tf-encrypted/issues/258

Checklist updated.

**Comment by mortendahl** [comment]
Turns out we don't really need share convert; skipping it only introduces a tiny error on the same scale as (either) truncation protocol. Let's keep code around for now in case some uses come up.

**Comment by ianlivingstone** [comment]
@mortendahl @bendecoste what's left for us to finish this issue?

**Comment by ianlivingstone** [comment]
Is this done @mortendahl ?

**Comment by mortendahl** [comment]
suggest we close this guy up, only some small improvements we could always make

**Comment by ianlivingstone** [comment]
@mortendahl awesome; if we know what those are, can you open new issues? I'll go ahead and close this one :)
